utxoscanner.go

```diff
@@ -0,0 +1,410 @@
+
+	// BlockFilterMatches checks the cfilter for the block hash for matches
+	// against the rescan options.
+	BlockFilterMatches func(ro *rescanOptions, blockHash *chainhash.Hash) (bool, error)
+
+	// GetBlock fetches a block from the p2p network.
+	GetBlock func(chainhash.Hash, ...QueryOption) (*btcutil.Block, error)
+}
+
+// UtxoScanner batches calls to GetUtxo so that a single scan can search for
+// multiple outpoints. If a scan is in progress when a new element is added, we
+// check whether it can safely be added to the current batch, if not it will be
+// included in the next batch.
+type UtxoScanner struct {
+	started uint32
+	stopped uint32
+
+	cfg *UtxoScannerConfig
+
+	pq        GetUtxoRequestPQ
+	nextBatch []*GetUtxoRequest
+
+	mu sync.Mutex
+	cv *sync.Cond
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+}
+
+// NewUtxoScanner creates a new instance of UtxoScanner using the given chain
+// interface.
+func NewUtxoScanner(cfg *UtxoScannerConfig) *UtxoScanner {
+	scanner := &UtxoScanner{
+		cfg:  cfg,
+		quit: make(chan struct{}),
+	}
+	scanner.cv = sync.NewCond(&scanner.mu)
+
+	return scanner
+}
+
+// Start begins running scan batches.
+func (s *UtxoScanner) Start() error {
+	if !atomic.CompareAndSwapUint32(&s.started, 0, 1) {
+		return nil
+	}
+
+	s.wg.Add(1)
+	go s.batchManager()
+
+	return nil
+}
+
+// Stop any in-progress scan.
+func (s *UtxoScanner) Stop() error {
+	if !atomic.CompareAndSwapUint32(&s.stopped, 0, 1) {
+		return nil
+	}
+
+	close(s.quit)
+	s.cv.Signal()
+
+	s.wg.Wait()
+
+	return nil
+}
+
+// Enqueue takes a GetUtxoRequest and adds it to the next applicable batch.
+func (s *UtxoScanner) Enqueue(outPoint *wire.OutPoint,
+	startHeight uint32) (*GetUtxoRequest, error) {
+
+	log.Debugf("Enqueuing request for %s with start height %d",
+		outPoint.String(), startHeight)
+
+	req := &GetUtxoRequest{
+		OutPoint:      outPoint,
+		StartHeight:   startHeight,
+		resultErrChan: make(chan reportAndError, 1),
+		quit:          s.quit,
+	}
+
+	s.cv.L.Lock()
+	select {
+	case <-s.quit:
+		s.cv.L.Unlock()
+		return nil, ErrShuttingDown
+	default:
+	}
+
+	// Insert the request into the queue and signal any threads that might be
+	// waiting for new elements.
+	heap.Push(&s.pq, req)
+
+	s.cv.L.Unlock()
+	s.cv.Signal()
+
+	return req, nil
+}
+
+// batchManager is responsible for scheduling batches of UTXOs to scan. Any
+// incoming requests whose start height has already been passed will be added to
+// the next batch, which gets scheduled after the current batch finishes.
+//
+// NOTE: This method MUST be spawned as a goroutine.
+func (s *UtxoScanner) batchManager() {
+	defer s.wg.Done()
+
+	for {
+		s.cv.L.Lock()
+		// Wait for the queue to be non-empty.
+		for s.pq.IsEmpty() {
+			s.cv.Wait()
+
+			select {
+			case <-s.quit:
+				s.cv.L.Unlock()
+				return
+			default:
+			}
+		}
+
+		// Re-queue previously skipped requests for next batch.
+		for _, request := range s.nextBatch {
```

Comment:
seems like it could be a problem if a batch is being processed, and only one new request comes in and is added to the `nextBatch`. When the batch processing is over, we won't start the next batch because the pq is empty?

Solution would be to just move this loop before the `IsEmpty` check above I think

Reply:
Ahh i see what you're saying, would making the condition:
`for s.pq.IsEmpty() && len(s.nextBatch) == 0 {`
solve this?

Reply:
Yeah, I think so. Or moving the loop pushing elements from `nextBatch` to the heap, as mentioned.

Reply:
And I think the latter is cleanest tbh 😬

Reply:
If we move the loop before, i think we would have the same issue. Say `nextBatch` is empty when we try to copy elements, and we only wait for the pq to be non-empty. Elements can be added to `nextBatch` without us recognizing.

I think the condition on the pq _and_ `nextBatch` w/ loop after handles these cases though

Reply:
I don't think this should be a problem, as elements are added to `nextBatch` only during ongoing scans, which is not happening at this point.

Reply:
Ahh you're right! I moved the nextbatch loop to the top as you suggested